daemon/logger/jsonfilelog/read.go

```diff
@@ -121,96 +119,106 @@ func tailFile(f io.ReadSeeker, logWatcher *logger.LogWatcher, tail int, since ti
 	}
 }
 
-func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
-	dec := json.NewDecoder(f)
-	l := &jsonlog.JSONLog{}
+func (l *JSONFileLogger) followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
+	var (
+		msg        *logger.Message
+		err        error
+		rotated    bool
+		readIndex  uint64
+		writeIndex uint64
+
+		dec        = json.NewDecoder(f)
+		log        = &jsonlog.JSONLog{}
+		subscriber = l.writeNotifier.Subscribe()
+		watchClose = logWatcher.WatchClose()
+	)
+
+	go func() {
+		for {
+			if _, ok := <-subscriber; !ok {
```

Comment:
The consumer might be slow sometimes. If we use the subscriber chan directly it may block the `Publish` or the publish may timeout.
The `writeIndex` is used for record how many times that the `Log` function is called. The `readIndex` is used for record how many lines have been consumed. This way the `Publish` never blocks, regardless of the speed of the consumer.